mkosi/__init__.py

```diff
@@ -385,9 +386,68 @@ def configure_os_release(context: Context) -> None:
 
             newosrelease.rename(osrelease)
 
+        osrelease_content = read_env_file(osrelease)
+
         if ArtifactOutput.os_release in context.config.split_artifacts:
             shutil.copy(osrelease, context.staging / context.config.output_split_os_release)
 
+    # https://www.freedesktop.org/software/appstream/docs/sect-Metadata-OS.html
+    osid = context.config.appstream_id or osrelease_content.get("ID", "linux")
+    name = context.config.appstream_name or osrelease_content.get("NAME") or context.config.image_id or osid
+    if context.config.appstream_description:
+        description = context.config.appstream_description
+    elif "PRETTY_NAME" in osrelease_content:
+        description = f"{osrelease_content.get("PRETTY_NAME")} built with mkosi"
+    else:
+        description = "Image built with mkosi"
+    icon = context.config.appstream_icon or "https://brand.systemd.io/assets/svg/systemd-logomark.svg"
+    if context.config.appstream_homepage:
+        home_url = context.config.appstream_homepage
+    elif "HOME_URL" in osrelease_content:
+        home_url = osrelease_content.get("HOME_URL")
+    else:
+        home_url = None
+    if home_url:
+        url = urlparse(home_url)
+        if not url.netloc:
+            home_url = None
+    if context.config.appstream_id:
+        id = context.config.appstream_id
+    elif home_url:
+        netloc = url.netloc.split(".")
+        netloc.reverse()
+        netloc.remove("www")
+        id = ".".join(netloc)
+        id += f".{osid}"
+    else:
+        id = osid
```

Comment:
This handles `home_url` in a single branch and hopefully also addresses @septatrix comment.
```suggestion
    if context.config.appstream_id:
        id = context.config.appstream_id
    elif home_url:
        url = urlparse(home_url)
        if url.netloc:
            netloc = url.netloc.removeprefix("www.").split(".")
            id = ".".join(reversed(netloc)) + f".{osid}"
        else:
            home_url = None
            id = osid
    else:
        id = osid
```

Reply:
it's the same result, doesn't make any difference

Reply:
I think having the logic for `home_url` in a single place is cleaner and removing the leading `www.` with `removeprefix` is shorter, and I'd argue cleaner, with the added check whether `www` is actually in there, that is needed by `remove`.

That being said, why do you want to strip the `www` in the first place? Above you said it's not a valid tld or domain and that is true, but neither is the fallback `linux` and in the reverse url scheme, it would end up at the end, just before `osid`, anyway. Also, this is only a fallback path, people should set the appstream id in the first place (maybe using specifiers), and singling out `www` seems a bit odd. 

I generally agree that it wouldn't make sense to have `www` in an appstream ID, but neither would other subdomains like `download`, as when I grab [current systemd](https://download.opensuse.org/repositories/system:/systemd/Debian_Testing/amd64/systemd_257.999+1648+g8e0037fb3-0_amd64.deb) from OBS. Can we do without stripping `www` or what is the logic here that I don't see?

@septatrix `list.remove` only removes the first occurence, so this also behaves the same. I misremembered this as well.

Reply:
> Can we do without stripping `www` or what is the logic here that I don't see?

This is from the spec:

> This means the reversed-DNS HOME_URL value from /etc/os-release (without any https/www parts) combined with the ID of the operating system will produce the operating system's component id.

---

> @septatrix `list.remove` only removes the first occurence, so this also behaves the same. I misremembered this as well

No they do not behave the same.
It removes the first occurrence, not the occurrence iff it is at index 0. Therefore a "www" subdomain somewhere in the middle of a URL would also be stripped which I am pretty sure is not how the spec is supposed to be interpreted (because then it would likely also mean removing any "www" component, not only the first occurrence.

Reply:
Thanks @septatrix, I really shouldn't look at this before having coffee :)